### PR TITLE
[CLI] [Next.js] Fix for jss create command overriding GraphQL edge endpoint

### PR DESCRIPTION
**Pull request by ambrauer**
## Description
* Added "name replacement mode" to applyNameToProject function (via optional "replaceName" parameter)
* Use new replaceName parameter for applyNameToProject in Next.js sample
* "graphqlEndpointPath" > "graphQLEndpointPath" name change in Next.js sample

## Motivation
* This fixes issues where a template may not follow the conventions. In particular, the Next.js template uses "/sitecore/api/graph/edge" for its graphQLEndpointPath (vs "/api/[app-name]")
* This also prevents the "sitecoreDistPath" from blindly being added to package.json (Next.js sample doesn't have one). This will only be updated with created app name if present.

## How Has This Been Tested?
Locally creating new nextjs projects with `jss create`. Ran Next.js sample app in all modes locally.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update (non-breaking change; modified files are limited to the `/docs` directory)

## Checklist:

- [ ] I have read the Contributing guide.
- [ ] My code follows the code style of this project.
- [ ] My code/comments/docs fully adhere to the Code of Conduct.
- [ ] My change is a code change and it requires an update to the documentation.
- [ ] My change is a documentation change and it requires an update to the navigation.

### CODE REVIEW

File: packages/sitecore-jss-cli/src/create/index.ts [hunk 1]
```diff
@@ -3,73 +3,144 @@ import fs from 'fs';
...
-  const packagePath = path.join(projectFolder, 'package.json');
+/**
+ * @param {PackageJson} pkg package.json object
+ * @param {string} name App name
+ * @param {string} [replaceName] Optional token which will enable "name replacement mode". If omitted, default/conventional values are used.
+ */
+export function applyNameToPackageJson(pkg: PackageJson, name: string, replaceName?: string) {
+  pkg.name = name; // root name will never match "replaceName", so always set directly
+  pkg.config.appName = replaceName
+    ? pkg.config.appName.replace(RegExp(escapeRegex(replaceName), 'g'), name)
```
Comment (illiakovalenko) [inline packages/sitecore-jss-cli/src/create/index.ts:33]: @ambrauer Can you define a separate func?
```js
const applyNameReplacement = (string value, string replaceName) => value.replace(RegExp(escapeRegex(replaceName), 'g'), name)
```
And replace all the same entries by it

**Review by illiakovalenko — CHANGES_REQUESTED** [review]

File: packages/sitecore-jss-cli/src/create/index.ts [hunk 2]
```diff
@@ -3,73 +3,144 @@ import fs from 'fs';
...
+  return literal.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
+}
 
-  const packagePath = path.join(projectFolder, 'package.json');
+/**
+ * @param {PackageJson} pkg package.json object
+ * @param {string} name App name
+ * @param {string} [replaceName] Optional token which will enable "name replacement mode". If omitted, default/conventional values are used.
+ */
+export function applyNameToPackageJson(pkg: PackageJson, name: string, replaceName?: string) {
```
Comment (anastasiya29) [inline packages/sitecore-jss-cli/src/create/index.ts:30]: Looking at this from the perspective of someone new to JSS, it's not obvious how we expect people to use "name replacement mode". Do we show example of valid `replaceName` values somewhere? I don't remember seeing this in docs. Maybe a backlog item for docs is all that's needed. I think adding an example of how the transform works to the function docs would be helpful too.

File: packages/sitecore-jss-cli/src/create/index.ts (hunk 1 above)
Comment (anastasiya29) [inline packages/sitecore-jss-cli/src/create/index.ts:33]: It seems unusual to initiate a RegEx with something that just ran through a string replace, as RegEx can perform replacements on matches, and include those in its result. Is it possible to let RegEx handle the escaping?

**Review by anastasiya29 — COMMENTED** [review]
Really great job on the unit test coverage

File: samples/nextjs/package.json [hunk 3]
```diff
@@ -10,7 +10,7 @@
     ],
     "sitecoreDistPath": "/dist/JssNextWeb",
     "sitecoreConfigPath": "/App_Config/Include/zzz",
-    "graphqlEndpointPath": "/sitecore/api/graph/edge",
+    "graphQLEndpointPath": "/sitecore/api/graph/edge",
```
Comment (anastasiya29) [inline samples/nextjs/package.json:13]: Right now, we figure out the graphQL endpoint by concatenating the Sitecore API hostname and this relative graphQLEndpointPath. Would there ever be a case when someone would need to use an absolute path for 'graphQLEndpointPath'?  For example, what about using both, SSR and SSG for a hybrid app - is it still a single endpoint that's smart enough to forward certain requests to the CD, or would we need to support using 2 endpoints?

File: samples/nextjs/scripts/generate-config.ts [hunk 4]
```diff
@@ -35,8 +35,8 @@ export function generateConfig(configOverrides?: { [key: string]: string }): voi
   // The GraphQL endpoint is an example of making a _computed_ config setting
   // based on other config settings.
   const computedConfig: { [key: string]: string } = {};
-  computedConfig.graphqlEndpoint =
-    '`${config.sitecoreApiHost}${config.graphqlEndpointPath}?sc_apikey=${config.sitecoreApiKey}`';
+  computedConfig.graphQLEndpoint =
+    '`${config.sitecoreApiHost}${config.graphQLEndpointPath}?sc_apikey=${config.sitecoreApiKey}`';
```
Comment (anastasiya29) [inline samples/nextjs/scripts/generate-config.ts:39]: Probably out of scope for this ticket, but I think we should capture Sitecore's query string param names into constants.

File: packages/sitecore-jss-cli/src/create/index.ts (hunk 1 above)
Comment (ambrauer) [reply packages/sitecore-jss-cli/src/create/index.ts]: @illiakovalenko Agree, refactored to dedicated function and added tests.

**Review by illiakovalenko — APPROVED** [review]

File: packages/sitecore-jss-cli/src/create/index.ts (hunk 1 above)
Comment (ambrauer) [reply packages/sitecore-jss-cli/src/create/index.ts]: @anastasiya29 The escape is there to make sure any special regex characters in the replaceName token are treated as literals in the search (and need to specifically use RegExp as first param). Illia had a comment about this bit, and has been refactored. I think it's more clear in the latest update, specifically the [related test](https://github.com/Sitecore/jss/blob/5c37cc7415f428166b13a97685e73a8acf5de906/packages/sitecore-jss-cli/src/create/index.test.ts#L40).

File: samples/nextjs/scripts/generate-config.ts (hunk 4 above)
Comment (ambrauer) [reply samples/nextjs/scripts/generate-config.ts:39]: Agree, would be a good future enhancement (looks like all the frameworks do this) but out of scope here.

File: samples/nextjs/package.json (hunk 3 above)
Comment (ambrauer) [reply samples/nextjs/package.json:13]: @anastasiya29 We may have to get more sophisticated with the `graphQLEndpoint` generation once we implement Exp Edge integration for sure. However, I'm not sure why the endpoint would differ between SSR / SSG. Either way, seems out of scope of this task.

File: packages/sitecore-jss-cli/src/create/index.ts (hunk 2 above)
Comment (ambrauer) [reply packages/sitecore-jss-cli/src/create/index.ts:36]: @anastasiya29 Agree, this isn't the clearest. Struggled to find a balance between this and previous functionality to maintain backwards compatibility. The `replaceName` argument is new with this PR, so not documented anywhere - though I don't think the whole `jss-create.js` and `applyNameToProject` were ever documented that I can find. I'll add a backlog item to document this better.

**Pull request merged**